pkg/api/errors/state.go

```diff
@@ -0,0 +1,114 @@
+		"state store is not configured",
+		"ERR_STATE_STORE_NOT_CONFIGURED",
+	).
+		WithErrorInfo(kiterrors.CodePrefixStateStore+kiterrors.CodeNotConfigured, nil).
+		Build()
+}
+
+func StateStoreNotFound(storeName string) error {
+	return kiterrors.NewBuilder(
+		codes.InvalidArgument, // TODO We should change this at some point. It should be codes.NotFound, but it will be a breaking change.
+		http.StatusBadRequest,
+		fmt.Sprintf("state store %s is not found", storeName),
+		"ERR_STATE_STORE_NOT_FOUND",
+	).
+		WithErrorInfo(kiterrors.CodePrefixStateStore+kiterrors.CodeNotFound, nil).
+		Build()
+}
+
+func StateStoreInvalidKeyName(storeName string, key string, msg string) error {
+	return kiterrors.NewBuilder(
+		codes.InvalidArgument,
+		http.StatusBadRequest,
+		msg,
+		"ERR_MALFORMED_REQUEST",
+	).WithErrorInfo(kiterrors.CodePrefixStateStore+kiterrors.CodeIllegalKey, nil).
+		WithResourceInfo("state", storeName, "", "").
+		WithFieldViolation(key, msg).
+		Build()
+}
+
+/**** Transactions ****/
+
+func StateStoreTransactionsNotSupported(storeName string) error {
+	return kiterrors.NewBuilder(
+		codes.Unimplemented,
+		http.StatusInternalServerError,
+		fmt.Sprintf("state store %s doesn't support transaction", storeName),
+		"ERR_STATE_STORE_NOT_SUPPORTED", // TODO: @elena-kolevska this code misleading and also used for different things ("query unsupported"); it should be removed in the next major version
+	).
+		WithErrorInfo(kiterrors.CodePrefixStateStore+"TRANSACTIONS_NOT_SUPPORTED", nil).
+		WithResourceInfo("state", storeName, "", "").
+		WithHelpLink("https://docs.dapr.io/reference/components-reference/supported-state-stores/", "Check the list of state stores and the features they support").
+		Build()
+}
+
+func StateStoreTooManyTransactionalOps(storeName string, count int, max int) error {
+	return kiterrors.NewBuilder(
+		codes.InvalidArgument,
+		http.StatusBadRequest,
+		fmt.Sprintf("the transaction contains %d operations, which is more than what the state store supports: %d", count, max),
+		"ERR_STATE_STORE_TOO_MANY_TRANSACTIONS",
+	).
+		WithErrorInfo(kiterrors.CodePrefixStateStore+"TOO_MANY_TRANSACTIONS", map[string]string{
+			"currentOpsTransaction": strconv.Itoa(count),
+			"maxOpsPerTransaction":  strconv.Itoa(max),
+		}).
+		WithResourceInfo("state", storeName, "", "").
+		Build()
+}
+
+/**** Query API ****/
+
+func StateStoreQueryUnsupported(storeName string) error {
+	return kiterrors.NewBuilder(
+		codes.Internal,
+		http.StatusInternalServerError,
+		"state store does not support querying",
+		"ERR_STATE_STORE_NOT_SUPPORTED",
+	).
+		WithErrorInfo(kiterrors.CodePrefixStateStore+"QUERYING_"+kiterrors.CodeNotSupported, nil).
+		WithResourceInfo("state", storeName, "", "").
+		Build()
+}
+
+func StateStoreQueryFailed(storeName string, detail string) error {
+	return kiterrors.NewBuilder(
+		codes.Internal,
+		http.StatusInternalServerError,
+		fmt.Sprintf("state store %s query failed: %s", storeName, detail),
```

Comment:
Do we need the name of the state store? It should be in the resource info so probably redundant here.
```suggestion
		fmt.Sprintf("query failed: %s", storeName, detail),
```

Reply:
That makes sense. Personally, I prefer the redundancy here though, because existing applications that haven't implemented the new error handling will still get meaningful error messages.